### PR TITLE
[AMDGPU][NewPM] Port SIOptimizeVGPRLiveRange pass to NPM.

### PR DESCRIPTION
**Pull request by cdevadas**

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIOptimizeVGPRLiveRange.cpp [hunk 1]
```diff
@@ -611,35 +620,62 @@ void SIOptimizeVGPRLiveRange::optimizeWaterfallLiveRange(
...
 
-bool SIOptimizeVGPRLiveRange::runOnMachineFunction(MachineFunction &MF) {
+PreservedAnalyses
+SIOptimizeVGPRLiveRangePass::run(MachineFunction &MF,
+                                 MachineFunctionAnalysisManager &MFAM) {
+  MFPropsModifier _(*this, MF);
+
+  if (MF.getFunction().hasOptNone())
+    return PreservedAnalyses::all();
+
```
Comment (optimisan) [inline llvm/lib/Target/AMDGPU/SIOptimizeVGPRLiveRange.cpp:657]: This is no longer required as OptNoneInstrumentation checks for it now.
https://github.com/llvm/llvm-project/blob/e914d97327ce5887e51c2d650987f2f48eda85c1/llvm/lib/Passes/StandardInstrumentations.cpp#L1050-L1060

File: llvm/lib/Target/AMDGPU/SIOptimizeVGPRLiveRange.cpp [hunk 2]
```diff
@@ -611,35 +620,62 @@ void SIOptimizeVGPRLiveRange::optimizeWaterfallLiveRange(
...
+  MachineLoopInfo *Loops = &MFAM.getResult<MachineLoopAnalysis>(MF);
 
+  bool Changed = SIOptimizeVGPRLiveRange(LV, MDT, Loops).run(MF);
+  if (!Changed)
+    return PreservedAnalyses::all();
+
+  auto PA = getMachineFunctionPassPreservedAnalyses();
+  PA.preserve<LiveVariablesAnalysis>();
+  PA.preserve<DominatorTreeAnalysis>();
+  PA.preserve<MachineLoopAnalysis>();
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIOptimizeVGPRLiveRange.cpp:669]: all CFG?

Comment (cdevadas) [reply llvm/lib/Target/AMDGPU/SIOptimizeVGPRLiveRange.cpp:669]: > all CFG?

Hmm. Not sure why that's not done in the Legacy path.

File: llvm/lib/Target/AMDGPU/SIOptimizeVGPRLiveRange.cpp [hunk 3]
```diff
@@ -611,35 +620,59 @@ void SIOptimizeVGPRLiveRange::optimizeWaterfallLiveRange(
...
+
+  bool Changed = SIOptimizeVGPRLiveRange(LV, MDT, Loops).run(MF);
+  if (!Changed)
+    return PreservedAnalyses::all();
+
+  auto PA = getMachineFunctionPassPreservedAnalyses();
+  PA.preserve<LiveVariablesAnalysis>();
+  PA.preserve<DominatorTreeAnalysis>();
+  PA.preserve<MachineLoopAnalysis>();
+  PA.preserveSet<CFGAnalyses>();
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIOptimizeVGPRLiveRange.cpp:666]: Should fix old pm in follow up

**Review by arsenm — APPROVED** [review]

**Comment by cdevadas** [comment]
### Merge activity

* **Nov 28, 10:38 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/117686).

**Pull request merged**